### PR TITLE
Midea support v2

### PR DESCRIPTION
**Pull request by dudanov**
# Midea support v2 (major update)

1. Most of code moved to separate independed library [dudanov/MideaUART](https://github.com/dudanov/MideaUART).
2. Component renamed from ``midea_ac`` to ``midea``.
3. Improved performance, responsiveness, feedback.
4. Added options for communicating with the device: minimum request period, response timeout, number of request attempts.
5. Added support for Midea IR protocol.
6. Automatic detection of air conditioner capabilities.
7. Added several actions for:
    - transmit IR FollowMe command;
    - toggle screen;
    - adjust the louver by one step;
    - control beeper feedback.
8. Settings options changed.
9. Some more related changes.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1356

## Test Environment

- [x] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
# Disable logging over UART (required)
logger:
  baud_rate: 0

# UART settings for Midea dongle (required)
uart:
  tx_pin: 1         # hardware dependant
  rx_pin: 3         # hardware dependant
  baud_rate: 9600

# Main settings
climate:
  - platform: midea
    name: Midea Climate         # Use a unique name.
    period: 1s
    timeout: 2s
    num_attempts: 3
    autoconf: true              # Autoconfigure most options.
    beeper: true                # Beep on commands.
    visual:                     # Optional. Example of visual settings override.
      min_temperature: 17 °C    # min: 17
      max_temperature: 30 °C    # max: 30
      temperature_step: 0.5 °C  # min: 0.5
    supported_modes:            # All capabilities in this section detected by autoconf.
      - FAN_ONLY                # This capability is always used.
      - HEAT_COOL
      - COOL
      - HEAT
      - DRY
    custom_fan_modes:
      - SILENT
      - TURBO
    supported_presets:          # All capabilities in this section detected by autoconf.
      - ECO
      - BOOST
      - SLEEP                   # This capability is always used.
    custom_presets:             # All capabilities in this section detected by autoconf.
      - FREEZE_PROTECTION
    supported_swing_modes:
      - VERTICAL                # This capability is always used.
      - HORIZONTAL
      - BOTH
    outdoor_temperature:        # Optional. Outdoor temperature sensor (may display incorrect values after long inactivity).
      name: Temp
    power_usage:                # Optional. Power usage sensor (only for devices that support this feature).
      name: Power
    humidity_setpoint:          # Optional. Indoor humidity sensor (only for devices that support this feature).
      name: Humidity

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by dudanov** [comment]
How do I fix the error of missing external library header files?

**Comment by jesserockz** [comment]
One thing to note, before I do an actual review, is can you add a `cv.invalid` to the `midea_ac` climate platform so users are not faced with `platform does not exist` but a nicer message saying it was renamed.

`esphome/components/midea_ac/climate.py`
```python
import esphome.config_validation as cv

CONFIG_SCHEMA = cv.invalid("This platform has been renamed to midea in 2021.9")
```

**Comment by dudanov** [comment]
> One thing to note, before I do an actual review, is can you add a `cv.invalid` to the `midea_ac` climate platform so users are not faced with `platform does not exist` but a nicer message saying it was renamed.
> 
> `esphome/components/midea_ac/climate.py`
> 
> ```python
> import esphome.config_validation as cv
> 
> CONFIG_SCHEMA = cv.invalid("This platform has been renamed to midea in 2021.9")
> ```
Done.

File: tests/test1.yaml [hunk 1]
```diff
@@ -1606,29 +1606,84 @@ climate:
...
     power_usage:
-      name: 'Power'
+      name: "Power"
     humidity_setpoint:
-      name: 'Hum'
+      name: "Humidity"
   - platform: anova
     name: Anova cooker
     ble_client_id: ble_blah
     unit_of_measurement: c
```
Comment (dudanov) [inline tests/test1.yaml:1651]: Here are the tests. Why absence label?

File: esphome/components/midea/climate.py [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+    if CONF_OUTDOOR_TEMPERATURE in config:
+        sens = await sensor.new_sensor(config[CONF_OUTDOOR_TEMPERATURE])
+        cg.add(var.set_outdoor_temperature_sensor(sens))
+    if CONF_POWER_USAGE in config:
+        sens = await sensor.new_sensor(config[CONF_POWER_USAGE])
+        cg.add(var.set_power_sensor(sens))
+    if CONF_HUMIDITY_SETPOINT in config:
+        sens = await sensor.new_sensor(config[CONF_HUMIDITY_SETPOINT])
+        cg.add(var.set_humidity_setpoint_sensor(sens))
+    cg.add_library("dudanov/MideaUART", "^1.1.0")
```
Comment (jesserockz) [inline esphome/components/midea/climate.py]: Would prefer if this was not a caret target
```suggestion
    cg.add_library("dudanov/MideaUART", "1.1.0")
```

Comment (dudanov) [reply esphome/components/midea/climate.py]: Done

**Review by jesserockz — APPROVED** [review]

**Pull request merged**